### PR TITLE
A another step towards library

### LINKED ISSUE #37: Library leftover
TODOs:
- [x] refactor resolver by taking submodules into consideration
- [x] reuse compiled core to speed up compilation
- [x] make recompile decision with a file dependency graph
- [ ] #216

/cc aya-prover/aya-prover-proto#652

### PR DESCRIPTION
**Pull request by imkiva**
This PR implements a really working library system. close #37

- refactor resolver by taking submodules into consideration
- reuse compiled core to speed up compilation
- make recompile decision with a file dependency graph
- serialize/deserialize `AyaBinOpSet`
- make operators scoped

### CODE REVIEW

**Comment by ice1000** [comment]
Suggestion: migrate the success tests to make it a package

File: base/src/main/java/org/aya/core/serde/SerDef.java [hunk 1]
```diff
@@ -30,8 +32,9 @@ record Fn(
     @NotNull SerTerm result
   ) implements SerDef {
     @Override public @NotNull Def de(SerTerm.@NotNull DeState state) {
+      DefVar<FnDef, Decl.FnDecl> ref = state.def(name);
       return new FnDef(
-        state.def(name), telescope.map(tele -> tele.de(state)),
+        ref, telescope.map(tele -> tele.de(state)),
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/serde/SerDef.java:37]: Why this change?

File: base/src/main/java/org/aya/tyck/order/IncrementalTycker.java [hunk 2]
```diff
@@ -7,20 +7,24 @@
 import org.aya.concrete.resolve.ResolveInfo;
 import org.aya.concrete.stmt.Decl;
 import org.aya.concrete.stmt.Stmt;
+import org.aya.util.MutableGraph;
 import org.jetbrains.annotations.NotNull;
 
 /**
  * Incremental and non-stopping compiler for SCCs.
  *
+ * @param declUsage   transpose of {@link ResolveInfo#declGraph()}. Vertex should be tycked first.
```
Comment (ice1000) [inline base/src/main/java/org/aya/tyck/order/IncrementalTycker.java:16]: Did you mean "Vertex should be tycked first"? Like, which vertices should be tycked first?

File: cli/src/main/java/org/aya/cli/library/LibraryModuleLoader.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ */
+public record LibraryModuleLoader(
+  @NotNull Reporter reporter,
+  @NotNull SourceFileLocator locator,
+  @NotNull SeqView<Path> thisModulePath,
+  @NotNull Path thisOutRoot
+) implements ModuleLoader {
+  public static @NotNull Path resolveCompiledCore(@NotNull Path basePath, @NotNull Seq<@NotNull String> moduleName) {
+    var withoutExt = moduleName.foldLeft(basePath, Path::resolve);
+    return withoutExt.resolveSibling(withoutExt.getFileName() + ".ayac");
```
Comment (ice1000) [inline cli/src/main/java/org/aya/cli/library/LibraryModuleLoader.java:49]: image

Can we merge these `+ "c"` things into one?

File: cli/src/main/java/org/aya/cli/library/CompiledAya.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    var modName = ctx.moduleName();
+    var exports = DynamicSeq.<SerDef.QName>create();
+    ctx.exports.view().forEach((k, vs) -> {
+      var qnameMod = modName.appendedAll(k);
+      vs.view().forEach((n, v) -> {
+        var qname = new SerDef.QName(qnameMod, n);
+        exports.append(qname);
+      });
+    });
```
Comment (ice1000) [inline cli/src/main/java/org/aya/cli/library/CompiledAya.java:39]: Use a `flatMap`?

File: cli/src/main/java/org/aya/cli/library/LibraryCompiler.java [hunk 5]
```diff
@@ -47,116 +75,222 @@
...
+    var source = new LibrarySource(owner, path);
+    var finder = new ImportResolver(owner, source);
+    finder.resolveStmt(program);
+    return source;
+  }
+
+  private @NotNull MutableGraph<LibrarySource> resolveLibraryImports() throws IOException {
+    var graph = MutableGraph.<LibrarySource>create();
+    var sources = collectSource(library.librarySrcRoot());
+    reporter.reportString("  [Info] Resolving source file dependency");
```
Comment (ice1000) [inline cli/src/main/java/org/aya/cli/library/LibraryCompiler.java:92]: Why this indentation?

Comment (ice1000) [reply cli/src/main/java/org/aya/cli/library/LibraryCompiler.java:92]: I also see other `reportString`s indented, but I wonder how the overall output would look like.

File: cli/src/main/java/org/aya/cli/Main.java [hunk 6]
```diff
@@ -24,28 +24,33 @@
 
   @Override
   public Integer call() throws Exception {
-    if (action == null) return 0;
+    if (action == null) {
+      System.err.println("Try `aya --help` to see available commands");
+      return 0;
```
Comment (ice1000) [inline cli/src/main/java/org/aya/cli/Main.java:29]: `return 0`? Maybe `return 1`?

**Review by ice1000 — COMMENTED** [review]
If there's a resolution or type error, will there still be highlighting?

**Comment by ice1000** [comment]
My reviews are mostly questions, so if you believe I didn't understand anything, just lmk

File: base/src/main/java/org/aya/tyck/order/IncrementalTycker.java (hunk 2 above)
Comment (imkiva) [reply base/src/main/java/org/aya/tyck/order/IncrementalTycker.java:16]: This is a usage graph: (vertex, w) means `w` depends on `vertex`.

File: cli/src/main/java/org/aya/cli/library/LibraryCompiler.java (hunk 5 above)
Comment (imkiva) [reply cli/src/main/java/org/aya/cli/library/LibraryCompiler.java:92]: image

File: cli/src/main/java/org/aya/cli/Main.java (hunk 6 above)
Comment (imkiva) [reply cli/src/main/java/org/aya/cli/Main.java:29]: Oh yes!

File: base/src/main/java/org/aya/core/serde/SerDef.java (hunk 1 above)
Comment (imkiva) [reply base/src/main/java/org/aya/core/serde/SerDef.java:37]: I am not sure whether the arguments are evaluated from left to right or right to left or unspecified.

Comment (ice1000) [reply base/src/main/java/org/aya/core/serde/SerDef.java:37]: Does it matter?

File: cli/src/main/java/org/aya/cli/library/LibraryCompiler.java (hunk 5 above)
Comment (ice1000) [reply cli/src/main/java/org/aya/cli/library/LibraryCompiler.java:92]: I see. Could these outputs possibly nest? If so, probably you wanna use `Doc`

File: base/src/main/java/org/aya/tyck/order/IncrementalTycker.java (hunk 2 above)
Comment (ice1000) [reply base/src/main/java/org/aya/tyck/order/IncrementalTycker.java:16]: Can you write "For each element (vertex, w) in the graph, the vertex should be tycked first" instead? I think the original comment is pretty vague

Comment (imkiva) [reply base/src/main/java/org/aya/tyck/order/IncrementalTycker.java:16]: Good idea

File: base/src/main/java/org/aya/core/serde/SerDef.java (hunk 1 above)
Comment (imkiva) [reply base/src/main/java/org/aya/core/serde/SerDef.java:37]: I tried reverting these changes and nothing was broken. Thanks for pointing them out!

File: cli/src/main/java/org/aya/cli/library/CompiledAya.java [hunk 7]
```diff
@@ -60,8 +60,8 @@ private static void ser(
     @NotNull DynamicSeq<SerDef> serDefs,
     @NotNull DynamicSeq<SerDef.SerOp> serOps
   ) {
+    var state = new Serializer.State();
     defs.forEach(def -> {
-      var state = new Serializer.State();
```
Comment (imkiva) [inline cli/src/main/java/org/aya/cli/library/CompiledAya.java]: Is OK to reuse serialize?

Comment (ice1000) [reply cli/src/main/java/org/aya/cli/library/CompiledAya.java]: no

Comment (ice1000) [reply cli/src/main/java/org/aya/cli/library/CompiledAya.java]: It's not ok to not reuse the state

**Review by ice1000 — APPROVED** [review]
lgtm so far
